Blog on mpOpenAPI (#4359)
OpenLiberty/blogs: posts/2025-03-28-Building-Better-APIs:-Documenting-APIs-in-Open-Liberty-using-mpOpenAPI.adoc

---
= Building Better APIs: Documenting APIs in Open Liberty using MicroProfile OpenAPI
Sweetty P Devassy <https://github.com/sweettypdevassy>
:imagesdir: /
:url-prefix:
:url-about: /

As technology has evolved and our applications have become more complicated, so too have our APIs. We are now building a huge number of highly complicated APIs. Documenting them requires effort and is often seen as a tedious task. Furthermore, the complex nature of these APIs can make documentation even more challenging. But why does documentation of our APIs matter? API documentation serves as a means of sharing up-to-date information between development teams, acting as a sort of how-to manual for developers that are looking to build things with your API. But having a standardized way of documenting our APIs so that we can all easily understand other APIs and what they do was a challenge... at least until the development of link:https://swagger.io/[Swagger] and later the link:https://www.openapis.org/[OpenAPI] specification. Having a structured approach to documentation helps developers improve collaboration, reduce errors, and accelerate development cycles.

> So, let's dive into the OpenAPI specification and explore how you can use it in your Open Liberty project with link:https://microprofile.io/specifications/open-api/[MicroProfile OpenAPI] features to easily and effectively document your APIs..

Reviewer: ```suggestion
So, let's dive into the OpenAPI specification and explore how you can use it in your Open Liberty project with link:https://microprofile.io/specifications/open-api/[MicroProfile OpenAPI] features to easily and effectively document your APIs.
```

== What is OpenAPI?  

OpenAPI is a standardised specification for describing RESTful APIs in a machine-readable format (JSON or YAML). It allows developers to define, document, and share API structures in a way that is both human-readable and programmatically useful. At its core, an OpenAPI definition includes:  

* *Info* – API title, version, description  

* *Servers* – Base URLs of the API  

* *Paths* – Endpoints (/users, /orders)  

* *Methods* – GET, POST, PUT, DELETE  

* *Parameters* – Query, path, header, or body inputs  

* *Responses* – Expected status codes and return data  

* *Security* – Authentication methods (OAuth, API keys, etc.)  

The OpenAPI specification can be used together with link:https://swagger.io/tools/swagger-ui/[Swagger UI] to allow developers to visualize the APIs and test them interactively. The great part about this integration is that the UI is automatically generated from your OpenAPI specification.

== MicroProfile OpenAPI and Open Liberty  

> For Open Liberty developers, link:https://microprofile.io/specifications/open-api/[MicroProfile OpenAPI] library can be used to integrate and make use of the OpenAPI specification. As well as enabling the documentation standards offered through the OpenAPI specification, the MicroProfile OpenAPI also eliminates some of the need for additional manual effort by offering the ability to automatically generate OpenAPI-compliant documentation for JAX-RS services. Saving even more time and effort! Making use of MicroProfile OpenAPI, Open Liberty developers are able to maintain consistent, accurate, and up-to-date API documentation. 

Reviewer: ```suggestion
For Open Liberty developers, the link:https://microprofile.io/specifications/open-api/[MicroProfile OpenAPI] library can be used to integrate and make use of the OpenAPI specification. As well as enabling the documentation standards offered through the OpenAPI specification, MicroProfile OpenAPI also eliminates some of the need for additional manual effort by offering the ability to automatically generate OpenAPI-compliant documentation for JAX-RS services, saving even more time and effort! Making use of MicroProfile OpenAPI, Open Liberty developers are able to maintain consistent, accurate, and up-to-date API documentation. 
```

Additionally, MicroProfile OpenAPI can be used seamlessly with the Swagger UI to enhance the developer and user experience by providing interactive API documentation, allowing users to visualise endpoints, test API calls, and understand request/response structures in a more intuitive way. This removes the need for manually written API documentation, reducing maintenance overhead and minimising the risk of outdated or inconsistent documentation. 

In summary, MicroProfile OpenAPI streamlines API development in Open Liberty, making it easier for teams to build, expose, and manage RESTful services efficiently.   

== Adding MicroProfile OpenAPI to Open Liberty 

> . Set Up Java

Reviewer: ```suggestion
. Set up Java
```

*3 . API Operations*

These annotations should be used with REST resource methods.

* *`@Operation`* - Describes an API operation.  

* *`@APIResponse`* - Defines possible responses for an operation.  

* *`@ApiResponses`* - Groups multiple @APIResponse annotations.  

* *`@Content`* - Specifies the media type and structure of request/response bodies.  

* *`@Schema`* - Defines the properties of a model used in API requests and responses.

[source,  java]
----
@GET
@Path("/user/{id}")
@Produces(MediaType.APPLICATION_JSON)
@Tag(name = "Blog")
@Operation(summary = "Get User by ID", description = "Returns the user details for a given ID")
@APIResponses({
@APIResponse(responseCode = "200", description = "Successful Response", content = @Content(
    mediaType = MediaType.APPLICATION_JSON, schema = @Schema(implementation = User.class)
)),
@APIResponse(responseCode = "404", description = "User not found")})
public User getUser(
----
[.img_border_light]
image::img/blog/OPENAPI-5.png[API Operations ,width=70%,align="center"] 
[.img_border_light]
image::img/blog/OPENAPI-6.png[API Operations Response ,width=70%,align="center"] 

*4 . Request Parameter Definitions*  

> This annotation should be used with REST resource method parameter.

Reviewer: ```suggestion
This annotation should be used with REST resource method parameters.
```

The MicroProfile OpenAPI feature is responsible for generating this documentation. This eliminates the need for external configuration files and ensures the documentation evolves alongside the codebase.

Ultimately, MicroProfile OpenAPI annotations play a vital role in creating useful, up-to-date documentation. Because they reside within the source code, they make it easier for teams to maintain alignment between implementation and documentation—benefiting both developers and end users who need to understand how to interact with the API effectively.

== Influence  

> As API architectures evolve, OpenAPI, the industry-standard specification for documenting REST APIs, is influencing how other tools and specification's APIs handle documentation and interoperability. Among those being influenced is link:https://graphql.org/learn/[GraphQL] and link:https://grpc.io/[gRPC]. GraphQL is a query language for your API, and a server-side runtime for executing queries using a type system you define for your data. gRPC is a modern open source high performance Remote Procedure Call (RPC) framework that can run in any environment. It can efficiently connect services in and across data centers with pluggable support for load balancing, tracing, health checking and authentication. While OpenAPI was designed for REST, its structured approach to API documentation, tooling, and automation has influenced how GraphQL and gRPC handle documentation and interoperability.

Reviewer: ```suggestion
As API architectures evolve, OpenAPI, the industry-standard specification for documenting REST APIs, is influencing how other tools' and specifications' APIs handle documentation and interoperability. Among those being influenced are link:https://graphql.org/learn/[GraphQL] and link:https://grpc.io/[gRPC]. GraphQL is a query language for your API, and a server-side runtime for executing queries using a type system you define for your data. gRPC is a modern open source high performance Remote Procedure Call (RPC) framework that can run in any environment. It can efficiently connect services in and across data centers with pluggable support for load balancing, tracing, health checking and authentication. While OpenAPI was designed for REST, its structured approach to API documentation, tooling, and automation has influenced how GraphQL and gRPC handle documentation and interoperability.
```
